niceman/resource/singularity.py

```diff
@@ -0,0 +1,240 @@
+
+        Returns
+        -------
+        dict : config parameters to capture in the inventory file
+        """
+        # Check to see if the container is already running.
+        if self.get_instance_info():
+            lgr.info('Resource {} already exists.'.format(self.name))
+        else:
+            self.image_file = "{}.simg".format(str(uuid.uuid4().hex)[:16])
+            if self.image.startswith('shub://') or \
+                    self.image.startswith('docker://'):
+
+                self._runner.run(['singularity', 'pull', '--name',
+                    self.image_file, self.image])
+            else:
+                shutil.copyfile(self.image, self.image_file)
+
+            # Start the container instance.
+            self._runner.run(['singularity', 'instance.start',
+                self.image_file, self.name])
+
+        # Update status
+        self.id = self.name
+        self.status = 'running'
+        return {
+            'id': self.id,
+            'status': self.status,
+            'image_file': self.image_file
+        }
+
+    def delete(self):
+        """
+        Deletes a container instance.
+        """
+        # Check to see if the container is already stopped.
+        if not self.get_instance_info():
+            return
+
+        # Stop the container.
+        self._runner.run(['singularity', 'instance.stop', self.name])
+
+        # Update status
+        self.id = None
+        self.status = None
+
+        # Remove temporary image file
+        os.remove(self.image_file)
+
+    def start(self):
+        """
+        Start a stopped container.
+        """
+        lgr.info("Singularity does not provide a start feature.")
+
+    def stop(self):
+        """
+        Stop a running container.
+        """
+        lgr.info("Singularity does not provide a stop feature.")
+
+    @borrowdoc(Resource)
+    def get_session(self, pty=False, shared=None):
+        if not self.get_instance_info():
+            self.connect()
+
+        if pty and shared is not None and not shared:
+            lgr.warning("Cannot do non-shared pty session for Singularity yet")
+        return (PTYSingularitySession if pty else SingularitySession)(
+            name=self.name
+        )
+
+    def get_instance_info(self):
+        """
+        Return basic information about a running container
+
+        Returns
+        -------
+        dict : instance info
+        """
+        try:
+            stdout, _ = self._runner.run(['singularity', 'instance.list'])
+        except CommandError:
+            return None
+
+        # Parse stdout to find the running instance.
+        # The output of instance.list is a table with columns:
+        #   DAEMON NAME, PID, and CONTAINER IMAGE
+        # Daemon names are unique on each server.
+        rows = stdout.split('\n')
+        for row in rows:
+            items = row.split()
+            if len(items) != 3:
+                # Skip header and trailing blank lines.
```

Comment:
With the s/split('\n')/splitlines/ change above, you could remove this check and just start the iteration at the second item.

Reply:
Yes, I can use slicing [1:]. Much more pythony!